### DISCUSSION TITLE
Usage an "await useFetch" or "await useAsyncData" in layouts

### DISCUSSION THREAD

**Discussion by ruslankonev** [nuxt/nuxt#18223] [Questions]
Using both `await useFetch` or `await useAsyncData` in the layouts will refresh the entire page and reloading all scripts. 
Is this the correct logic to work and my wrong way of using `<script setup>` or is there something wrong?

**Comment by wobsoriano** [comment]
Do you have a repro?

**Reply by ruslankonev** [reply]
@wobsoriano repro link → https://codesandbox.io/s/mystifying-taussig-omjdt 
In browser view → https://omjdt.sse.codesandbox.io/

**Reply by atinux** [reply]
It should not reload the page, we will investigate, do you mind opening a bug report?

If this cannot be fixed, it shall be in the documentation.

Thank you for your help on debugging Nuxt 😊

**Comment by wobsoriano** [comment]
Based on the [docs](https://v3.nuxtjs.org/docs/usage/data-fetching#useasyncdata), I think `useFetch` and `useAsyncData` are only meant for pages, components, and plugins that's why a full page reload happens when navigating if you put it in a `layout`.

I tried to use it in a plugin to set initial value like this:

```ts
// plugins/state.ts
export default defineNuxtPlugin(async () => {
    const { data } = await useAsyncData("mainpage", () =>
        $fetch("https://jsonplaceholder.typicode.com/todos")
    );

    const data = useState("mainData", () => data);
})
```

But I got this error:

```
Cannot read properties of null (reading '_nuxtOnBeforeMountCbs')
```

So I tried manually fetching and setting the value:

```ts
export default defineNuxtPlugin(() => {
    const data = useState("mainData", () => []);

    $fetch("https://jsonplaceholder.typicode.com/todos").then((value) => {
        data.value = value
    })
})
```

Which worked.

Not sure if the error above is a bug.

**Reply by wobsoriano** [reply]
https://github.com/nuxt/framework/issues/1344

**Reply by ruslankonev** [reply]
Some context. Refreshing the entire page, will reproduce in layouts and `app.vue` only if `useFetch` or `useAsyncData` was used with the `await` keyword before

**Reply by ruslankonev** [reply]
And some related question with usage of `<script setup>` in layouts I describe [there](https://github.com/nuxt/framework/discussions/1342) → it always called when page changed https://github.com/nuxt/framework/discussions/1342

**Reply by oleg-kazannikov-mediapark** [reply]
It's worked for me
```js
export default defineNuxtPlugin(async () => {
    const data = useState("mainData", () => []);

    data.value = await $fetch("https://jsonplaceholder.typicode.com/todos");
})
```
